### PR TITLE
Mock xhr-fetch

### LINKED ISSUE #10477: Support XHR mocking in describes.js
- Add a dead-simple way to stub/mock `Window.fetch` when using describes.

Related: https://github.com/ampproject/amphtml/issues/10128#issuecomment-311496211 and #9338

Internal discussion: `https://groups.google.com/a/google.com/d/msg/amphtml-eng/YKxMVlpQ6xA/fj3KS3FDAQAJ`

/cc @lannka @taymonbeal @jridgewell @rsimha-amp

### PR DESCRIPTION
**Pull request by Raphaeljunior**
- Mocks out fetch in realWin and fakeWin and avails it through expectFetch as follows.

Closes #10477

### CODE REVIEW

**Comment by rsimha-amp** [comment]
@Raphaeljunior, once #10671 is merged, the dependency errors on Travis should no longer be an issue.

File: package.json [hunk 1]
```diff
@@ -48,6 +48,7 @@
     "eslint": "3.18.0",
     "eslint-plugin-google-camelcase": "0.0.2",
     "express": "4.14.0",
+    "fetch-mock": "^5.12.1",
```
Comment (rsimha-amp) [inline package.json]: Can you also update `yarn.lock` by running `yarn upgrade --save-dev` and including the diffs in this PR?

/cc @erwinmombay

Comment (jridgewell) [inline package.json]: Don't use `yarn upgrade`, use `yarn install`.

Comment (Raphaeljunior) [inline package.json]: It's an error that happens on travis.

Comment (jridgewell) [inline package.json]: Because you upgraded the `yarn.lock` to use unsupported versions. Revert changes to `yarn.lock`, then run `yarn install` and it'll work again.

**Comment by Raphaeljunior** [comment]
@jridgewell PR ready for review.

File: test/functional/test-xhrMock.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview
+ * Asserts xhr fetch has been mocked out and under control
+ **/
```
Comment (choumx) [inline test/functional/test-xhrMock.js]: Nit: Single `*`.

```js
/**
 *
 */
```

File: test/functional/test-xhrMock.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
```
Comment (choumx) [inline test/functional/test-xhrMock.js]: 2017

File: testing/describes.js [hunk 4]
```diff
@@ -112,7 +117,7 @@ import {resetLoadingCheckForTests} from '../src/element-stub';
 import {resetScheduledElementForTesting} from '../src/custom-element';
 import {setStyles} from '../src/style';
 import * as sinon from 'sinon';
-
+import fetchMock from 'fetch-mock';
```
Comment (choumx) [inline testing/describes.js]: Nit: Insert newline after this.

File: testing/describes.js [hunk 5]
```diff
@@ -187,7 +192,8 @@ export const sandboxed = describeEnv(spec => []);
  * A test with a fake window.
  * @param {string} name
  * @param {{
- *   win: !FakeWindowSpec,
+ *   win: !FakeWindow
+Spec,
```
Comment (choumx) [inline testing/describes.js]: Revert.

File: testing/describes.js [hunk 6]
```diff
@@ -341,7 +347,8 @@ function describeEnv(factory) {
 
   /**
    * @param {string} name
-   * @param {!Object} spec
+   * @param {!Object} specconsole.log(fetchMock.sandbox());
+          debugger;
```
Comment (choumx) [inline testing/describes.js]: Revert.

File: testing/describes.js [hunk 7]
```diff
@@ -503,6 +517,7 @@ class RealWinFixture {
   constructor(spec) {
     /** @const */
     this.spec = spec;
+    // throw spec.toString();
```
Comment (choumx) [inline testing/describes.js]: Revert.

File: testing/describes.js [hunk 8]
```diff
@@ -483,15 +487,25 @@ class FakeWinFixture {
 
   /** @override */
   setup(env) {
+    const spec = this.spec;
     env.win = new FakeWindow(this.spec.win || {});
+    if (!(spec.xhrMock === false)) {
```
Comment (choumx) [inline testing/describes.js]: `if (!!this.spec.xhrMock) {`

File: testing/describes.js [hunk 9]
```diff
@@ -483,15 +487,25 @@ class FakeWinFixture {
 
   /** @override */
   setup(env) {
+    const spec = this.spec;
```
Comment (choumx) [inline testing/describes.js]: Either update the following line to also use this local var or remove this.

File: testing/describes.js [hunk 10]
```diff
@@ -556,7 +571,13 @@ class RealWinFixture {
         interceptEventListeners(win.document.documentElement);
         interceptEventListeners(win.document.body);
         env.interceptEventListeners = interceptEventListeners;
-
+        if (!(spec.xhrMock === false)) {
+          fetchMock.constructor.global = env.win;
+          fetchMock._mock();
+          env.expectFetch = function(url, response) {
+            return fetchMock.mock(url, response);
+          };
```
Comment (choumx) [inline testing/describes.js]: `env.expectFetch = fetchMock.mock.bind(fetchMock)`?

File: .travis.yml [hunk 11]
```diff
@@ -2,7 +2,7 @@ language: node_js
 sudo: required  # See http://docs.travis-ci.com/user/trusty-ci-environment/
 dist: trusty
 node_js:
-  - "4"
+  - "6"
```
Comment (choumx) [inline .travis.yml]: Do we still need to upgrade to Node 6? I think this was resolved, right?

File: testing/describes.js [hunk 12]
```diff
@@ -556,7 +571,13 @@ class RealWinFixture {
...
+        }
```
Comment (choumx) [inline testing/describes.js]: Let's move this block of code to a top-level helper function to avoid code dupe.

File: build-system/tasks/presubmit-checks.js [hunk 13]
```diff
@@ -74,7 +74,7 @@ var forbiddenTerms = {
   'sinon\\.(spy|stub|mock)\\(': {
     message: 'Use a sandbox instead to avoid repeated `#restore` calls',
   },
-  '(\\w*([sS]py|[sS]tub|[mM]ock|clock).restore)': {
+  '(\\w*([sS]py|[sS]tub|\\.[mM]ock|clock).restore)': {
```
Comment (choumx) [inline build-system/tasks/presubmit-checks.js]: Why this change?

File: extensions/amp-a4a/0.1/test/fetch-mock.js [hunk 14]
```diff
@@ -1,4 +1,4 @@
-/**
+  /**
```
Comment (choumx) [inline extensions/amp-a4a/0.1/test/fetch-mock.js]: Revert.

File: extensions/amp-ad/0.1/test/test-amp-ad-xorigin-iframe-handler.js [hunk 15]
```diff
@@ -161,6 +161,7 @@ describe('amp-ad-xorigin-iframe-handler', () => {
         return initPromise.then(() => {
           expect(noContentSpy).to.be.calledWith(false);
           expect(iframeHandler.iframe).to.be.null;
+          done();
```
Comment (choumx) [inline extensions/amp-ad/0.1/test/test-amp-ad-xorigin-iframe-handler.js]: This shouldn't be necessary since we're returning a Promise in the test.

File: extensions/amp-bind/0.1/test/integration/test-bind-impl.js [hunk 16]
```diff
@@ -109,6 +109,7 @@ describe.configure().skipSauceLabs().run('Bind', function() {
       ampdoc: 'fie',
       runtimeOn: false,
     },
+    xhrMock: false,
```
Comment (choumx) [inline extensions/amp-bind/0.1/test/integration/test-bind-impl.js]: Nit: How about `mockFetch`? Sounds more like an action/verb.

File: build-system/tasks/presubmit-checks.js (hunk 13 above)
Comment (Raphaeljunior) [inline build-system/tasks/presubmit-checks.js]: To prevent the presubmit.js from blocking fetchMock.restore()

**Comment by choumx** [comment]
@taymonbeal FYI, let us know if this works for you.

**Comment by choumx** [comment]
Now that @Raphaeljunior has left, closing this and forking into #11001 since it's a huge pain to patch with conflicts and ~80 commits.
